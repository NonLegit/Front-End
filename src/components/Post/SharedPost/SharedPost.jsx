/* eslint-disable react/no-danger */
// mui components
import {
  Box, useMediaQuery, useTheme,
} from '@mui/material';
import ArrowForwardIosIcon from '@mui/icons-material/ArrowForwardIos';
import ArrowBackIosNewIcon from '@mui/icons-material/ArrowBackIosNew';

// styles
import { useRef, useEffect, useState } from 'react';

import { useNavigate } from 'react-router-dom';
import {
  PostContainer, PostMedia, CustomImage, PostText, PostTextContainer, ControlsIcon, PostUrl, PostUrlLink, LinkIcon,

} from './styles';

import PostHeader from './PostHeader/PostHeader';
/**
 * This component is the view of the post in home page.
 *
 * @component Post
 * @property {Post} sharedFrom -The parent post of the current one.
 * @property {boolean} subreddit -to identify if post in home page or subreddit.
 * @returns {React.Component} Post
 */

function Post(props) {
  const { sharedFrom, subredit } = props;
  const temp = {
    _id: 1,
    text: '',
    backgroundColor: '',
    textColor: '',
  };
  const {
    text: flairText,
    backgroundColor: flairBackgroundColor,
    textColor: flairColor,
  } = (sharedFrom?.flairId) || temp;
  const { name: ownerName, icon: ownerIcon } = sharedFrom?.owner || {};
  const { name: authorName } = sharedFrom?.author || {};

  console.log('sharedfrom', sharedFrom);
  const {
    _id: postId, createdAt, title, images, kind, text, video, ownerType, url, nsfw, spoiler,
  } = sharedFrom;

  console.log('wow', {
    _id: postId, createdAt, title, images, ownerName, ownerIcon, authorName, flairText, flairBackgroundColor, flairColor, kind, text, video, ownerType, url, nsfw, spoiler,
  });
  // routes
  const navigate = useNavigate();

  // styles
  const theme = useTheme();
  const matchSm = useMediaQuery(theme.breakpoints.up('sm'));

  // variables
  const maxTextHeight = 180;

  // states
  const [displayShadow, setDisplayShadow] = useState(false);
  const [index, setIndex] = useState(0);
  const [maxImagesHeight, setMaxImagesHeight] = useState(450);

  // refs
  const postTextRef = useRef();
  const postMediaRef = useRef();

  // handlers
  const handleDirection = (dir) => {
    setIndex(index + dir);
  };

  const redirectToPost = (redirect) => {
    if (redirect) {
      const username = ownerName;
      if (ownerType === 'User') {
        if (username) {
          navigate(`/user/${username}/comments/${postId}`);
        } else {
          navigate('/');
        }
      } else {
        navigate(`/r/${ownerName}/comments/${postId}`);
      }
    }
  };

  // effects
  useEffect(() => {
    setDisplayShadow(postTextRef?.current?.offsetHeight > maxTextHeight);
  }, [text]);

  useEffect(() => {
    images?.forEach((image) => {
      const img = new Image();

      img.src = image;
      img.onload = () => {
        setMaxImagesHeight((maxImagesHeight) => {
          const maxValue = Math.min(maxImagesHeight, img.height);
          const postWidth = postMediaRef?.current?.offsetWidth;
          if (maxImagesHeight > img.height && img.width > postWidth) {
            return img.height * (postWidth / img.width);
          }
          return maxValue;
        });
      };
    });
  }, [images]);

  // console.log('for post ', postId, maxImagesHeight);
  return (
    <PostContainer>
      <Box
        p={1}
        flexGrow={1}
        maxWidth={matchSm ? '94.5%' : '100%'}
      >
        <PostHeader
          title={title}
          ownerIcon={ownerIcon}
          ownerName={ownerName}
          authorName={authorName}
          flairText={flairText}
          flairBackgroundColor={flairBackgroundColor}
          flairColor={flairColor}
          createdAt={createdAt}
          subredit={subredit}
          ownerType={ownerType}
          nsfw={nsfw}
          spoiler={spoiler}
          redirectToPost={redirectToPost}
        />
        {/* eslint-disable jsx-a11y/media-has-caption */}
        {/* */}
        {!nsfw && (
        <PostMedia
          mt={1.5}
          kind={kind}
          ref={postMediaRef}
          spoiler={spoiler}
          onClick={() => redirectToPost(kind !== 'link')}
        >
          {kind === 'video' ? (
            <video controls style={{ width: '100%', maxHeight: '512px' }}>
              <source src={video} type="video/mp4" />
            </video>
          ) : (
            (kind === 'image')
              ? (
                <>
                  {images.map((image, imageIndex) => (
                    imageIndex === index
                    && (
                    <CustomImage
                      src={image}
                      alt="post image"
                      key={image}
                      maxHeight={maxImagesHeight}
                    />
                    )
                  ))}
                  <>
                    <ControlsIcon
                      disableRipple
                      left={10}
                      display={index <= 0 ? 'none' : 'flex'}
                      sx={{
                        boxShadow: 10,
                      }}
                      color="third"
                      onClick={(e) => {
                        e.stopPropagation();
                        handleDirection(-1);
                      }}
                    >
                      <ArrowBackIosNewIcon />
                    </ControlsIcon>
                    <ControlsIcon
                      disableRipple
                      right={10}
                      display={index >= images.length - 1 ? 'none' : 'flex'}
                      sx={{
                        boxShadow: 10,
                      }}
                      color="third"
                      onClick={(e) => {
                        e.stopPropagation();
                        handleDirection(1);
                      }}
                    >
                      <ArrowForwardIosIcon />
                    </ControlsIcon>
                  </>
                </>
              ) : ((kind === 'self') ? (
                <PostText
                  ref={postTextRef}
                  maxHeight={displayShadow ? maxTextHeight : 'none'}
                >
                  {displayShadow && <PostTextContainer />}
                  <div dangerouslySetInnerHTML={{ __html: text }} />
                </PostText>
              )
                : (
                  <PostUrl href={url}>
                    <PostUrlLink>
                      {url}
                    </PostUrlLink>
                    <LinkIcon />
                  </PostUrl>
                ))
          )}
        </PostMedia>
        )}
      </Box>
    </PostContainer>
  );
}

export default Post;
