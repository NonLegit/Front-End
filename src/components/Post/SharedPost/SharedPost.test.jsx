import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { ThemeProvider } from '@mui/system';
import { MemoryRouter as Router } from 'react-router-dom';
import theme from '../../../styles/theme';
import SharedPost from './SharedPost';

describe(SharedPost, () => {
  it('SharedPost renders correctly', () => {
    const testCase = JSON.parse(`{
        "_id": "63a039107e9602e5da79ac42",
        "ownerType": "User",
        "replies": [
            "63a03c427e9602e5da79ac8b",
            "63a03c457e9602e5da79ac97",
            "63a03c4a7e9602e5da79aca3",
            "63a03c4c7e9602e5da79acaf"
        ],
        "title": "new Post @ 12:12",
        "kind": "self",
        "text": "<p>Edited Post</p>",
        "url": "",
        "images": [],
        "createdAt": "2022-12-19T10:06:05.781Z",
        "locked": false,
        "isDeleted": false,
        "sendReplies": true,
        "nsfw": false,
        "spoiler": false,
        "votes": 0,
        "views": 0,
        "commentCount": 4,
        "shareCount": 0,
        "suggestedSort": "new",
        "scheduled": false,
        "modState": "unmoderated",
        "spamCount": 0,
        "spammedBy": [],
        "sortOnHot": 835722182890.5,
        "sortOnBest": 668577746312.4,
        "__v": 0,
        "owner": {
            "_id": "639f824cd9f48a8d75a21aa4",
            "name": "Basma",
            "icon": "https://api.nonlegit.click/users/default.png"
        },
        "author": {
            "_id": "639f824cd9f48a8d75a21aa4",
            "name": "Basma"
        },
        "isSaved": false,
        "postVoteStatus": 0,
        "isSpam": false
      }`);
    const tree = renderer
      .create(
        <Router>
          <ThemeProvider theme={theme}>
            <SharedPost
              sharedFrom={testCase}
              subredit
            />
          </ThemeProvider>
        </Router>
        ,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
