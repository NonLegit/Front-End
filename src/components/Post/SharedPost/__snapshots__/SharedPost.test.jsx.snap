// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post SharedPost renders correctly 1`] = `
<div
  className="MuiBox-root css-1va79tx"
>
  <div
    className="MuiBox-root css-1roq6yo"
  >
    <div
      className="MuiBox-root css-1eogohz"
    >
      <div
        className="MuiBox-root css-1clzuz0"
      >
        <div
          className="MuiBox-root css-5ah3mn"
        >
          4 days ago
        </div>
      </div>
    </div>
    <a
      className="css-10m9jmh"
      href="/"
      onClick={[Function]}
    >
      <h3
        className="MuiTypography-root MuiTypography-h6 css-6aghek-MuiTypography-root"
      >
        new Post @ 12:12
         
         
      </h3>
    </a>
    <div
      className="MuiBox-root css-vp8937"
      kind="self"
      onClick={[Function]}
      spoiler={false}
    >
      <p
        className="css-w0u38i"
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Edited Post</p>",
            }
          }
        />
      </p>
    </div>
  </div>
</div>
`;
